cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -92,55 +99,69 @@ void mergeToken(BasicWorkflowToken other) {
 
   @Override
   public void put(String key, String value) {
+    put(key, value, Scope.USER);
+  }
+
+  @Override
+  public void put(String key, Value value) {
+    put(key, value.toString());
+  }
+
+  void put(String key, String value, Scope scope) {
     if (nodeName == null) {
       throw new IllegalStateException("Node name cannot be null.");
     }
-    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
+
+    List<NodeValueEntry> nodeValueList = tokenValueMap.get(scope, key);
     if (nodeValueList == null) {
       nodeValueList = Lists.newArrayList();
-      tokenValueMap.put(key, nodeValueList);
+      tokenValueMap.put(scope, key, nodeValueList);
     }
 
     // Check if the current node already added the key to the token.
     // In that case replace that entry with the new one
     for (int i = 0; i < nodeValueList.size(); i++) {
       if (nodeValueList.get(i).getNodeName().equals(nodeName)) {
-        nodeValueList.set(i, new NodeValueEntry(nodeName, value));
+        nodeValueList.set(i, new NodeValueEntry(nodeName, new Value(value)));
         return;
       }
     }
 
-    nodeValueList.add(new NodeValueEntry(nodeName, value));
+    nodeValueList.add(new NodeValueEntry(nodeName, new Value(value)));
   }
 
+  @Nullable
```

Comment:
If this method returns null, we should move the ``@Nullable`` annotation to the interface.